From b88fb0286f33dc92829be379d7635a7e6836664a Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Thu, 23 Jan 2020 11:15:18 +0200
Subject: [PATCH] PONRTSYS-6525: 6rd traffic is not accelerated

Fix the IPv4 checksum calculation for all recipes for inner IPv6
and outer IPv4 packets

Signed-off-by: Emergi Idan <idan.emergi@intel.com>
---
 .../net/ppv4/pp-hal/modifier/modifier_recipes.c    | 66 ++++++++--------------
 1 file changed, 24 insertions(+), 42 deletions(-)

diff --git a/drivers/net/ppv4/pp-hal/modifier/modifier_recipes.c b/drivers/net/ppv4/pp-hal/modifier/modifier_recipes.c
index bc29baee4c27..56127d9392d2 100644
--- a/drivers/net/ppv4/pp-hal/modifier/modifier_recipes.c
+++ b/drivers/net/ppv4/pp-hal/modifier/modifier_recipes.c
@@ -531,10 +531,6 @@ static s32 mod_rcp_ipv4_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 	s32 pkt_l3_ttl         = DPU_RCP_REG4_B3;
 	s32 pkt_l4_csum        = DPU_RCP_REG5_B0;
 	s32 pkt_l3_tot_len     = DPU_RCP_REG6_B0;
-	/* Used to access ttl as a 16bit register required by csum update */
-	s32 tmp_ttl_16         = DPU_RCP_REG7_B0;
-	/* Copy of pkt_l3_ttl */
-	s32 tmp_ttl            = DPU_RCP_REG7_B1;
 	s32 si_l4_csum_zero    = DPU_RCP_REG8_B0;
 
 	/* USE registers R12 - R15 with extra caution as they are
@@ -616,12 +612,13 @@ static s32 mod_rcp_ipv4_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 	/*  |------------------------|------------------------------------|
 	 *  |      ALU0              |      ALU1                          |
 	 *  |------------------------|------------------------------------|
-	 *  | sync                   | sync (nhdr or pppoe_encap)         |
-	 *  | TTL decrement          | pkt l3 csum + si delta             |
-	 *  | ID inc (nhdr v4)       | nhdr tot len (nhdr or pppoe_encap) |
-	 *  | csum ID (nhdr v4)      | pppoe pkt len (pppoe_encap)        |
-	 *  | csum TTL (nhdr v4)     | pkt l4 csum step 1                 |
-	 *  | csum tot len (nhdr v4) | pkt l4 csum step 2                 |
+	 *  | sync                   |                                    |
+	 *  | TTL decrement          | sync                               |
+	 *  | ID inc (nhdr v4)       | pkt l3 csum + si delta             |
+	 *  | csum ID (nhdr v4)      | nhdr tot len (nhdr or pppoe_encap) |
+	 *  | csum tot len (nhdr v4) | pppoe pkt len (pppoe_encap)        |
+	 *  |                        | pkt l4 csum step 1                 |
+	 *  |                        | pkt l4 csum step 2                 |
 	 *  |------------------------|------------------------------------|
 	 */
 	/* Synchronize ALU0 to latest fetched information of STORE command */
@@ -632,7 +629,7 @@ static s32 mod_rcp_ipv4_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 
 	/* decrement TTL */
 	ret |= mod_alu_sub(EU_ALU0, &rcp, pkt_l3_ttl, si_l3_ttl_diff,
-			   tmp_ttl, FIELD_SIZEOF(struct iphdr, ttl));
+			   pkt_l3_ttl, FIELD_SIZEOF(struct iphdr, ttl));
 
 	/* Recalculate L3 checksum */
 	ret |= mod_alu_csum16_add(EU_ALU1, &rcp, si_l3_csum_delta, pkt_l3_csum,
@@ -661,12 +658,6 @@ static s32 mod_rcp_ipv4_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 		ret |= mod_alu_csum16_sub(EU_ALU0, &rcp, MOD_GLB_EXT_IPV4_ID,
 					  si_nhdr_csum, si_nhdr_csum);
 
-		/* Update the external L3 csum due to ttl,
-		 * csum update must be performed on 16 bit operands
-		 */
-		ret |= mod_alu_csum16_sub(EU_ALU0, &rcp, tmp_ttl_16,
-					  si_nhdr_csum, si_nhdr_csum);
-
 		/* Update the external L3 csum due to total len */
 		ret |= mod_alu_csum16_sub(EU_ALU0, &rcp, pkt_l3_tot_len,
 					  si_nhdr_csum, si_nhdr_csum);
@@ -701,7 +692,7 @@ static s32 mod_rcp_ipv4_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 				  FIELD_SIZEOF(struct iphdr, tos));
 
 	/* Replace L3 TTL */
-	ret |= mod_reg_to_hdr_cpy(&rcp, tmp_ttl, l3_off,
+	ret |= mod_reg_to_hdr_cpy(&rcp, pkt_l3_ttl, l3_off,
 				  offsetof(struct iphdr, ttl),
 				  FIELD_SIZEOF(struct iphdr, ttl));
 
@@ -815,10 +806,6 @@ static s32 mod_rcp_ipv6_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 	s32 pkt_hop_limit   = DPU_RCP_REG2_B0;
 	/* s32 pkt_nexthdr       = DPU_RCP_REG2_B1; */
 	s32 pkt_payload_len = DPU_RCP_REG2_B2;
-	/* Used to access ttl as a 16bit register required by csum update */
-	s32 tmp_ttl_16      = DPU_RCP_REG3_B0;
-	/* Copy of pkt_hop_limit */
-	s32 tmp_ttl         = DPU_RCP_REG3_B1;
 	s32 ext_l3_hdr_len  = DPU_RCP_REG3_B3;
 	s32 stw_ext_l3_off  = DPU_RCP_REG5_B0;
 	s32 stw_int_l3_off  = DPU_RCP_REG5_B1;
@@ -869,16 +856,16 @@ static s32 mod_rcp_ipv6_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 	/*=========================*/
 	/*      ALU operations     */
 	/*=========================*/
-	/*  |------------------------|---------------------------------------|
-	 *  |      ALU0              |      ALU1                             |
-	 *  |------------------------|---------------------------------------|
-	 *  | sync                   | sync (nhdr or pppoe_encap)            |
-	 *  | HOPL decrement         | nhdr tot len (nhdr or pppoe_encap)    |
-	 *  | ID inc (nhdr v4)       | pkt v6 hdr len (nhdr or pppoe_encap)  |
-	 *  | csum TTL (nhdr v4)     | nhdr pkt len (nhdr or pppoe_encap)    |
-	 *  | csum ID (nhdr v4)      | pppoe len (nhdr or pppoe_encap)       |
-	 *  | csum pkt len (nhdr v4) |                                       |
-	 *  |------------------------|---------------------------------------|
+	/*  |------------------------------- |------------------------------|
+	 *  |      ALU0                      |      ALU1                    |
+	 *  |------------------------------- |------------------------------|
+	 *  | sync                           |                              |
+	 *  | HOPL decrement                 | sync (nhdr or pppoe)         |
+	 *  | pkt v6 hdr len (nhdr or pppoe) | nhdr tot len (nhdr or pppoe) |
+	 *  | ID inc (nhdr v4)               | nhdr pkt len (nhdr or pppoe) |
+	 *  | csum ID (nhdr v4)              | pppoe len (nhdr or pppoe)    |
+	 *  | csum pkt len (nhdr v4)         |                              |
+	 *  |------------------------        |------------------------------|
 	 */
 	/* Synchronize ALU to latest fetched information of STORE command */
 	ret |= mod_alu_synch(EU_ALU0, &rcp, l3_off,
@@ -891,7 +878,8 @@ static s32 mod_rcp_ipv6_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 				     FIELD_SIZEOF(struct ipv6hdr, hop_limit));
 
 	/* decrement hop limit */
-	ret |= mod_alu_sub(EU_ALU0, &rcp, pkt_hop_limit, si_ttl_diff, tmp_ttl,
+	ret |= mod_alu_sub(EU_ALU0, &rcp, pkt_hop_limit, si_ttl_diff,
+			   pkt_hop_limit,
 			   FIELD_SIZEOF(struct ipv6hdr, hop_limit));
 
 	/* Calculate new header total and/or pppoe length */
@@ -901,7 +889,7 @@ static s32 mod_rcp_ipv6_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 				   pkt_payload_len, 2, pkt_payload_len);
 
 		/* calc the inner v6 header length */
-		ret |= mod_alu_sub(EU_ALU1, &rcp, stw_int_l4_off,
+		ret |= mod_alu_sub(EU_ALU0, &rcp, stw_int_l4_off,
 				   stw_int_l3_off, ext_l3_hdr_len, 1);
 
 		/* add the inner v6 header length */
@@ -916,18 +904,12 @@ static s32 mod_rcp_ipv6_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 						EU_ALU_COND_MET);
 	}
 
-	/* for V4 update csum, id, and ttl */
+	/* for V4 update csum, id */
 	if (nhdr == RCP_NEW_HDR_V4) {
 		/* Increment the global id */
 		ret |= mod_alu_inc(EU_ALU0, &rcp, MOD_GLB_EXT_IPV4_ID,
 				   MOD_GLB_EXT_IPV4_ID);
 
-		/* Update the external L3 csum due to ttl
-		 * csum update must be performed on 16 bit operands
-		 */
-		ret |= mod_alu_csum16_sub(EU_ALU0, &rcp, tmp_ttl_16,
-					  si_nhdr_csum, si_nhdr_csum);
-
 		/* Update the external L3 csum due to global id */
 		ret |= mod_alu_csum16_sub(EU_ALU0, &rcp, MOD_GLB_EXT_IPV4_ID,
 					si_nhdr_csum, si_nhdr_csum);
@@ -945,7 +927,7 @@ static s32 mod_rcp_ipv6_routed_set(s32 rcp_idx, enum mod_eu_stw_mux_sel l3_off,
 	/* copy from dpu registers to packet (reps) */
 	/*==========================================*/
 	/* Replace L3 hop limit */
-	ret |= mod_reg_to_hdr_cpy(&rcp, tmp_ttl, l3_off,
+	ret |= mod_reg_to_hdr_cpy(&rcp, pkt_hop_limit, l3_off,
 				  offsetof(struct ipv6hdr, hop_limit),
 				  FIELD_SIZEOF(struct ipv6hdr, hop_limit));
 
